src/platform/testing/e2e/form-tester/form-filler.js

```diff
@@ -195,14 +261,12 @@ const fillPage = async (page, testData, testConfig, log = () => {}) => {
       await enterData(page, field, findData(field.selector, testData), log);
     }
 
-    // TODO: Check for arrays and click the appropriate add buttons
-    // TODO: Check for multiple add buttons and warn if there are
-
-    newFieldCount = await page.$$eval(
-      FIELD_SELECTOR,
-      elements => elements.length,
-    );
-  } while (fieldCount !== newFieldCount);
+    // If we have newly-expanded fields, they may be array fields.
+    // Add a new array item as needed only after we have no more expanded fields.
+    if (snapshotsAreEqual(originalSnapshot, await getSnapshot(page))) {
+      await addNewArrayItem(page, testData);
+    }
+  } while (!snapshotsAreEqual(originalSnapshot, await getSnapshot(page)));
```

Comment:
We can no longer check to see if just the number of fields is the same as it was before. When we add a new array item, the number of fields is the same, but the fields for the new item sorta just take the place of the old ones. We now have to make sure that we didn't add a new array item as well.

If the comments in the code don't clearly reflect this, please let me know how I can improve them. My goal is to make it as self-documenting as possible.

Reply:
I think it would be easier to read if some of the callbacks used named functions. e.g. I don't know what this do while loop is doing with fields without reading the implementation. 

`snapshots` is a little confusing terminology- it might be better to call `snapshoptsAreEqual` to `pageHasNewFields`, `fieldCountChanged`, `hasEmptyFields`, or `fieldsNeedInput`. Something that more closely describes the behavior rather than describing how it's done.

Reply:
Yeah, good point.

I thought about `hasNewFields`, but thought that might get confusing. They technically _are_ new fields, but the `fieldCount` doesn't actually change. We go from
```
    Field list: [ 'query',
      'root_view:hasAlternateName',
      'root_view:hasAlternateName',
      'root_alternateNames_0_first',
      'root_alternateNames_0_middle',
      'root_alternateNames_0_last' ]
```
to
```
    Field list: [ 'query',
      'root_view:hasAlternateName',
      'root_view:hasAlternateName',
      'root_alternateNames_1_first',
      'root_alternateNames_1_middle',
      'root_alternateNames_1_last' ]
```

:thinking: So if the test data doesn't have information for all the fields on a page, we won't enter anything for the missing fields. This makes names like `hasEmptyFields` maybe a bit misleading. What name would indicate that we've got new fields that we still have test data for? I think I like `fieldsNeedInput`.